torch/csrc/jit/codegen/cuda/root_domain_map.cpp

```diff
@@ -0,0 +1,694 @@
+#include <torch/csrc/jit/codegen/cuda/root_domain_map.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+
+#include <sstream>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+std::unordered_map<IterDomain*, IterDomain*> RootDomainMap::
+    mapProducerToConsumer(
+        const TensorDomain* producer,
+        const TensorDomain* consumer,
+        const std::unordered_set<IterDomain*>& root_dims_to_map) const {
+  return map(producer, consumer, root_dims_to_map, true);
+}
+
+std::unordered_map<IterDomain*, IterDomain*> RootDomainMap::
+    mapConsumerToProducer(
+        const TensorDomain* consumer,
+        const TensorDomain* producer,
+        const std::unordered_set<IterDomain*>& root_dims_to_map) const {
+  return map(producer, consumer, root_dims_to_map, false);
+}
+
+PairwiseRootDomainMap::PairwiseRootDomainMap(
+    const TensorView* producer,
+    const TensorView* consumer)
+    : producer_tv_(producer), consumer_tv_(consumer) {
+  TORCH_INTERNAL_ASSERT(producer != nullptr);
+  TORCH_INTERNAL_ASSERT(consumer != nullptr);
+  // Make sure they are really a producer and its consumer
+  Expr* origin = consumer->getOrigin();
+  TORCH_INTERNAL_ASSERT(origin != nullptr);
+  const auto& producer_exprs = producer->fusion()->unordered_exprs();
+  TORCH_INTERNAL_ASSERT(
+      producer_exprs.find(origin) != producer_exprs.end(),
+      "Not a producer-consumer pair: ",
+      producer,
+      ", ",
+      consumer);
```

Comment:
What are you trying to assert above here? `fusion()->unordered_exprs()` will return all exprs in the fusion, they're not related to the producer except for the fact that they're in the same fusion. Couldn't you just do `producer->fusion() == consumer->fusion()`, then check the inputs of `consumer->getOrigin()` to see that one of them is producer?

Reply:
Thanks. That was my mistake. I think I was looking at `Fusion::unordered_uses(Val*)`, but somehow I ended up writing that code. Will fix it.